stairway/src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -55,130 +46,123 @@ public FlightMap() {
-
-    // TODO(PF-703): Remove this block.
-    if (!inputList.isEmpty()) {
-      try {
-        map.validateAgainst(inputList);
-      } catch (Exception ex) {
-        Logger logger = LoggerFactory.getLogger("FlightMap");
-        logger.error("Input list does not match JSON: {}", ex.getMessage());
-      }
+  static FlightMap create(List<FlightInput> inputList, @Nullable String json) {
+
+    // If we have an empty list AND there is valid JSON, this predates the flightworking table...
+    // use the json in this case.
+    if (inputList.isEmpty() && json != null) {
+      FlightMap map = new FlightMap();
+      map.fromJson(json);
+      return map;
     }
 
-    return Optional.of(map);
-  }
-
-  /**
-   * Validate whether the passed inputList will deserialize to the same set of keys and
-   * corresponding types stored in the map. Note that this does not ensure the equality of values as
-   * all types stored in FlightMap are not required to be comparable.
-   *
-   * <p>Throws {@link RuntimeException} if any key exists in inputList but not in map (or vice
-   * versa). Throws {@link JsonProcessingException} if any value in inputList does not deserialize
-   * to the type of its corresponding entry in the map.
-   *
-   * @param inputList
-   */
-  @VisibleForTesting
-  void validateAgainst(List<FlightInput> inputList) throws Exception {
-
-    if (inputList.size() != map.size())
-      throw new RuntimeException(
-          String.format(
-              "Passed input list has %d entries, map has %d.", inputList.size(), map.size()));
-
-    for (FlightInput input : inputList) {
-      final String key = input.getKey();
-      if (!map.containsKey(key)) {
-        throw new RuntimeException(String.format("Key '%s' not found in map.", key));
-      }
-      final Object object = map.get(key);
-      if (object != null) {
-        getObjectMapper().readValue(input.getValue(), object.getClass());
-      }
-    }
+    // Otherwise just use the input list (even if it's empty).
+    return new FlightMap(inputList);
   }
 
   /** Convert the map to an unmodifiable form. */
   public void makeImmutable() {
     map = Collections.unmodifiableMap(map);
   }
 
+  /** Check map for emptiness */
+  boolean isEmpty() {
+    return map.isEmpty();
+  }
+
   /**
-   * Return the object from the hash map cast to the right type. Throw an exception if the Object
-   * cannot be cast to that type.
+   * Return the object from the hash map deserialized to the right type. Throw an exception if the
+   * Object cannot be deserialized to that type.
    *
    * @param <T> - type of class to expect in the hash map
    * @param key - key to lookup in the hash map
    * @param type - class requested
    * @return null if not found
-   * @throws ClassCastException if found, not castable to the requested type
+   * @throws ClassCastException if found, not deserializable to the requested type
    */
+  @Nullable
   public <T> T get(String key, Class<T> type) {
```

Comment:
Now that you're calling ObjectMapper on a per-field basis, you could add an overloaded `get()` method that accepts `TypeReference<T>` instead of `Class<T>`. That would make it possible to deserialize generics with type safety, and users can rely on type inference instead of specifying the type as a parameter.

Another nice feature would be to return `Optional<T>` instead of `T` which would make the method return policy clearer. But that would require making a new API and deprecating this one.

Reply:
Thanks for this feedback...  I'd like to explore this a bit more to get a better feel for the benefits we can get from this, would it be OK if we continued this thread in a follow-on PR so I can start propagating the in-database changes downstream?

Reply:
> would it be OK if we continued this thread in a follow-on PR so I can start propagating the in-database changes downstream?
That's fine with me.

> I'd like to explore this a bit more to get a better feel for the benefits we can get from this
There's been a bunch of articles written about how this API is better than `Class<T>`, here's one. https://northcoder.com/post/jackson-object-mapper-which-way-is/